src/commands.ts

```diff
@@ -56,7 +56,14 @@ const digits = keyValueBy(Array(58 - 48).fill(0), (n, i) => ({
 
 /** Hash all the properties of a command into a string that can be compared with the result of hashKeyDown. */
 export const hashCommand = (command: Command): string => {
+  if (!command.keyboard) return ''
+
+  // If command.keyboard is an array, return empty string as we handle arrays in index()
+  if (Array.isArray(command.keyboard)) return ''
```

Comment:
This is a bit awkward. When you find yourself adding strange edge cases for things that are not directly related to the function, you should take a look at the overall design of the function and consider if it needs to change.

Here, `hashCommand` does not depend on anything other than `command.keyboard`. It was written to take `command: Command` out of convenience. But now that it's not convenient anymore, it's not a great approach. Passing the entire `Command` also makes handling keyboard arrays more awkward in `index()`.

Instead, change `hashCommand` to accept `Command['keyboard']` and the function will become simpler.

Reply:
Even when we just accept `Command['keyboard'] `type, don't we still need to handle `Key[]` type explicitly either by adding edge case like above or by using something like `Exclude<Command['keyboard'], Key[]>`?